### PR TITLE
refactor: ABFS implementation

### PR DESCRIPTION
**Pull request by majetideepak**
Combine AbfsAccount and AbfsConfig in a separate file.
Clean up API naming and clarify semantics.
Add a new constructor for AbfsWriteFile to specify a client. This is used for testing.

### CODE REVIEW

**Comment by FelixYBW** [comment]
@zhli1142015

File: velox/connectors/hive/storage_adapters/abfs/AdlsFileClient.h [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ * Azurite Simulator does not yet support the DFS endpoint.
+ * (For more information, see https://github.com/Azure/Azurite/issues/553 and
+ * https://github.com/Azure/Azurite/issues/409).
+ * You can find a comparison between DFS and Blob endpoints here:
+ * https://github.com/Azure/Azurite/wiki/ADLS-Gen2-Implementation-Guidance
+ *
+ * To facilitate unit testing of file write scenarios, we define the
+ * AzureDatalakeFileClient here, which can be mocked during testing.
+ */
+class AdlsFileClient {
```
Comment (zhli1142015) [inline velox/connectors/hive/storage_adapters/abfs/AdlsFileClient.h:38]: Please use AzureDatalakeFileClient, AdlsFileClient indicates a different thing in our context.
Thanks.

File: velox/connectors/hive/storage_adapters/abfs/tests/MockAdlsFileClient.h [hunk 2]
```diff
@@ -14,19 +14,29 @@
...
+namespace facebook::velox::filesystems {
+/*
+ * The Azurite Simulator does not yet support the DFS endpoint.
+ * (For more information, see https://github.com/Azure/Azurite/issues/553 and
+ * https://github.com/Azure/Azurite/issues/409).
+ * You can find a comparison between DFS and Blob endpoints here:
+ * https://github.com/Azure/Azurite/wiki/ADLS-Gen2-Implementation-Guidance
+ *
+ * To facilitate unit testing of file write scenarios, we define the
+ * IBlobStorageFileClient here, which can be mocked during testing.
```
Comment (zhli1142015) [inline velox/connectors/hive/storage_adapters/abfs/tests/MockAdlsFileClient.h:32]: Please update the name here also.

File: velox/connectors/hive/storage_adapters/abfs/AbfsConfig.h [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+#include <folly/hash/Hash.h>
+#include <string>
+
+namespace facebook::velox::config {
+class ConfigBase;
+}
+
+namespace facebook::velox::filesystems {
+
+static std::string kAzureBlobEndpoint{"fs.azure.blob-endpoint"};
```
Comment (zhli1142015) [inline velox/connectors/hive/storage_adapters/abfs/AbfsConfig.h:28]: What does this property do? Is this for testing only, if yes please add comments.
Thanks.

File: velox/connectors/hive/storage_adapters/abfs/AdlsFileClient.h (hunk 1 above)
Comment (majetideepak) [reply velox/connectors/hive/storage_adapters/abfs/AdlsFileClient.h]: The azure client name is `DataLakeFileClient`. I renamed the implementations accordingly and tried to keep the name short.

**Comment by majetideepak** [comment]
@zhli1142015 I noticed that the usage of DataLakeFileClient::Flush is not optimal. We flush when the file is closed
and this might consume a lot of memory if the file written is big. We need to flush similarly to S3 periodically (10Mib chunks). What do you think?

File: velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp [hunk 4]
```diff
@@ -126,25 +111,35 @@ class AbfsWriteFile::Impl {
...
 
 AbfsWriteFile::AbfsWriteFile(
-    const std::string& path,
-    const std::string& connectStr) {
-  impl_ = std::make_shared<Impl>(path, connectStr);
+    std::string_view path,
+    const config::ConfigBase& config) {
+  auto abfsAccount = AbfsConfig(path, config);
+  std::shared_ptr<AzureDataLakeFileClient> client =
+      std::make_shared<DataLakeFileClientWrapper>(
```
Comment (czentgr) [inline velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp:125]: Just curious why the DataLakeFileClientWrapper is shared_ptr and not unique_ptr? Would something else access this if it is part of the client?

File: velox/connectors/hive/storage_adapters/abfs/tests/AbfsCommonTest.cpp [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+      config);
+  EXPECT_EQ(abfssConfig.fileSystem(), "abc");
+  EXPECT_EQ(
+      abfssConfig.filePath(),
+      "sf_1/store_sales/ss_sold_date_sk=2450816/part-00002-a29c25f1-4638-494e-8428-a84f51dcea41.c000.snappy.parquet");
+  EXPECT_EQ(
+      abfssConfig.connectionString(),
+      "DefaultEndpointsProtocol=https;AccountName=foobar;AccountKey=456;EndpointSuffix=core.windows.net;");
+
+  // test with special characters
```
Comment (czentgr) [inline velox/connectors/hive/storage_adapters/abfs/tests/AbfsCommonTest.cpp:61]: Nit. Commit not complete sentence.

File: velox/connectors/hive/storage_adapters/abfs/tests/AzuriteServer.cpp [hunk 6]
```diff
@@ -15,15 +15,33 @@
...
+}
+
+std::shared_ptr<const config::ConfigBase> AzuriteServer::hiveConfig(
+    const std::unordered_map<std::string, std::string> configOverride) const {
+  auto endpoint = fmt::format("http://127.0.0.1:{}/{}", port_, account_);
+  std::unordered_map<std::string, std::string> config(
+      {{"fs.azure.account.key.test.dfs.core.windows.net", key_},
+       {kAzureBlobEndpoint, endpoint}});
+
+  // Update the default config map with the supplied configOverride map
```
Comment (czentgr) [inline velox/connectors/hive/storage_adapters/abfs/tests/AzuriteServer.cpp:39]: Nit. Missing `.`. Or do we even need this comment?

File: velox/connectors/hive/storage_adapters/abfs/tests/AzuriteServer.h [hunk 7]
```diff
@@ -46,19 +37,40 @@ class AzuriteServer {
...
   bool isRunning();
 
-  void addFile(std::string source, std::string destination);
+  void addFile(std::string source);
 
   virtual ~AzuriteServer();
 
  private:
   int64_t port_;
+  std::string account_{"test"};
```
Comment (czentgr) [inline velox/connectors/hive/storage_adapters/abfs/tests/AzuriteServer.h:66]: Maybe all of these 4 new members are const?

**Review by czentgr — COMMENTED** [review]
nice refactor. looks much more clean now.

**Comment by zhli1142015** [comment]
> ient::Flush is not optimal. We flush when the file is closed
> and this might consume a lot of memory if the file written is big. We need to flush similarly to S3 periodically (10Mib chunks). What do you think?

I thought we don't cache data and send by chunk. The `append` API sends data to remote directly.
The behavior you mentioned in S3, may be better, we can take a look later.
Thanks.

**Comment by majetideepak** [comment]
@czentgr thanks for the review. I addressed your comments.

File: velox/connectors/hive/storage_adapters/abfs/tests/AzuriteServer.cpp (hunk 6 above)
Comment (majetideepak) [reply velox/connectors/hive/storage_adapters/abfs/tests/AzuriteServer.cpp]: I updated the function documentation instead.

File: velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp (hunk 4 above)
Comment (majetideepak) [reply velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp]: shared_ptr made it easier to write tests. But I changed this to a unique_ptr as I think it should be as well.

**Comment by majetideepak** [comment]
Filed https://github.com/facebookincubator/velox/issues/11456 for the write improvements.

File: velox/connectors/hive/storage_adapters/abfs/tests/MockDataLakeFileClient.h [hunk 8]
```diff
@@ -14,26 +14,36 @@
...
-  MockBlobStorageFileClient() {
-    auto tempFile = ::exec::test::TempFilePath::create();
+  MockDataLakeFileClient() {
+    auto tempFile = velox::exec::test::TempFilePath::create();
     filePath_ = tempFile->getPath();
   }
 
+  MockDataLakeFileClient(std::string_view filePath) : filePath_(filePath) {}
+
+  std::string_view path() {
```
Comment (czentgr) [inline velox/connectors/hive/storage_adapters/abfs/tests/MockDataLakeFileClient.h:35]: Nit, this could be `std::string_view path() const`.

**Review by czentgr — APPROVED** [review]
Looks good. Just one nit.

File: velox/connectors/hive/storage_adapters/abfs/tests/MockDataLakeFileClient.h (hunk 8 above)
Comment (majetideepak) [reply velox/connectors/hive/storage_adapters/abfs/tests/MockDataLakeFileClient.h]: done!

**Comment by kevinwilfong** [comment]
This needs a maintainer to approve, cc: @Yuhta @xiaoxmeng 

https://velox-lib.io/docs/community/components-and-maintainers/

**Comment by kevinwilfong** [comment]
I just double checked with the PLC and it sounds like that's not sufficient.

**Comment by majetideepak** [comment]
@zhli1142015 Can you please take a look and approve this PR?

**Review by zhli1142015 — APPROVED** [review]

**Review by Yuhta — APPROVED** [review]
Stamp since @zhli1142015 has the most knowledge about this and he is good with the change.

**Comment by kevinwilfong** [comment]
It looks like if we merge this it will break Presto, could you make a VELOX_ENABLE_FORWARD_COMPATIBILITY change in Presto to handle changing the namespace

https://github.com/prestodb/presto/blob/ecfda83fe2987a191b1cd8f722e9fe2e9d4c8b0e/presto-native-execution/presto_cpp/main/PrestoServer.cpp#L1304

**Comment by majetideepak** [comment]
@kevinwilfong Added Presto change here https://github.com/prestodb/presto/pull/24063

**Comment by kevinwilfong** [comment]
Thanks! I'll try merging this again.

**Pull request closed without merging**